### PR TITLE
fix: resolve C# loader regressions (#3108, #3107, #3140, version-info) 

### LINKED ISSUE #3108: [Bug]: Startup scripts fail in 6.1
### ✈ Pre-Flight checks

- [x] I **don't have SentinelOne** antivirus installed (see above for the solution)
- [x] I have **searched in the issues** (open and closed) but couldn't find a similar issue
- [x] I have **searched in the [pyRevit Forum](https://discourse.pyrevitlabs.io)** for similar issues
- [x] I already **followed the [installation troubleshooting guide](https://pyrevitlabs.notion.site/Installation-issues-f6495adb56254e9e8021f8a68e3b2ab7) thoroughly**
- [x] I am using the **latest pyRevit Version**

### 🐞 Describe the bug

may have been in 6.0 already

worked without logging a warning in 5.3.1

### ⌨ Error/Debug Message

```shell
WARNING [pyrevit.loader.sessionmgr] Startup script returned non-zero result for extension: MyTools, result: 1
```

### ♻️ To Reproduce

_No response_

### ⏲️ Expected behavior

_No response_

### 🖥️ Hardware and Software Setup (please complete the following information)

```shell
master | Branch: "master" | Version: "6.1.0.26047+2255" | Path: "C:\Program Files\pyRevit-Master"
```

### Additional context

```python
import os
import subprocess
from pyrevit import forms, script, EXEC_PARAMS

logger = script.get_logger()
my_config = script.get_config("autoupdate")
if not my_config.get_option("autoupdate", True):
    script.exit()

basepath = os.path.dirname(EXEC_PARAMS.command_path)
command = os.path.join(basepath, "MyTools.extension", "dev", "GetMyToolsFromGit.bat")
BAT_COMMAND = "update"

if not os.path.exists(command):
    forms.alert(
        "Nothing done. Update script not found at:\n{}".format(command),
        title="Updater ERROR",
        exitscript=True,
    )


try:
    task = subprocess.Popen(
        [command, BAT_COMMAND], stdout=subprocess.PIPE, stderr=subprocess.PIPE
    )
    out, err = task.communicate()

    # Log output if present
    if out:
        output_text = out.decode("utf-8") if isinstance(out, bytes) else out
        logger.info("Script output: {}".format(output_text.strip()))

    # Log errors if present
    if err:
        error_text = err.decode("utf-8") if isinstance(err, bytes) else err
        logger.error("Script error: {}".format(error_text.strip()))

    # Check return code and provide feedback
    if task.returncode != 0:
        logger.error(
            "Update script failed with return code: {}".format(task.returncode)
        )
    else:
        logger.info("Update script completed successfully.")

except Exception as e:
    logger.error("Failed to execute update script: {}".format(str(e)))
```

### PR DESCRIPTION
**Pull request by tay0thman**
# resolve C# loader regressions 

## Description
- #3108: Restore multi-pass loop in SessionManagerService.LoadSession() so all assemblies load before any startup script runs. Fixes cross-extension imports in startup scripts.

- #3107: Prefix underscore when SanitizeClassName() produces a name starting with a digit (invalid C# identifier). Applied in ExtensionParser, CommandTypeGenerator, ButtonBuilderBase, StackBuilder.

- #3140: Add HashSet dedup guard in CommandTypeGenerator.GenerateExtensionCode() to skip duplicate class names instead of emitting CS0101 Roslyn errors.

- Version info: Pass revitYear through ExtensionManagerService to ExtensionParser.ParseInstalledExtensions() so min/max version filtering works correctly. Previously defaulted to 0 (unknown).

---

## Checklist

Before submitting your pull request, ensure the following requirements are met:

- [OK ] Changes are tested and verified to work as expected.

---

## Related Issues

If applicable, link the issues resolved by this pull request:

- Resolves #3108: , #3107, #3140: 

---

## Additional Notes

Tested on Revit 2026 (netcore) and 2024(netfx)

---

Thank you for contributing to pyRevit! 🎉

### CODE REVIEW

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/AssemblyMaker/CommandTypeGenerator.cs [hunk 1]
```diff
@@ -49,20 +49,41 @@ private static string GetPyRevitRoot()
...
+                    // Duplicate — skip this command to avoid CS0101.
+                    // Emit a comment in the generated .cs so the user can diagnose
+                    // which script was dropped (the .cs file is saved alongside the DLL).
+                    sb.AppendLine($"// WARNING [#3140]: Skipped duplicate class '{safeClassName}'");
+                    sb.AppendLine($"//   Script: {cmd.ScriptPath ?? "(no script)"}");
+                    sb.AppendLine($"//   UniqueId: {cmd.UniqueId}");
+                    sb.AppendLine($"//   Two bundle directories produced the same UniqueId.");
+                    sb.AppendLine($"//   Rename one directory to fix this.");
+                    sb.AppendLine();
+                    continue;
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/AssemblyMaker/CommandTypeGenerator.cs:82]: "The duplicate scenario means two directories collided to the same identifier. The first command wins; the second is logged as a warning. This matches legacy behavior where the first-encountered command took precedence. A full dedup-with-suffix approach is out of scope for this bugfix PR."

File: dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs [hunk 2]
```diff
@@ -79,6 +79,58 @@ private static void LogParseException(string parsedFile, Exception ex)
...
+                if (!int.TryParse(minRevitVersion, out var min))
+                {
+                    LogWarning($"'{name}': min_revit_version value '{minRevitVersion}' is not a valid integer - skipping.");
+                    compatible = false;
+                }
+                else if (revitYear < min)
+                {
+                    LogInfo($"'{name}': skipped - requires Revit {min} or later (running {revitYear}).");
+                    compatible = false;
+                }
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs:114]: being edited now

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/AssemblyMaker/CommandTypeGenerator.cs [hunk 3]
```diff
@@ -228,9 +265,8 @@ public static string BuildEngineConfigs(ParsedComponent cmd, string scriptPath)
             
             if (isDynamoScript)
             {
-                // For Dynamo scripts, use appropriate settings
-                // Use automate or mainthread setting (automate is Dynamo-specific synonym)
-                bool requiresMainThread = (cmd.Engine?.MainThread ?? false) || (cmd.Engine?.Automate ?? true);
+                // Use EngineConfig.RequiresMainThread which already has the correct defaults.
+                bool requiresMainThread = cmd.Engine?.RequiresMainThread ?? false;
                 configs["automate"] = requiresMainThread;
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/AssemblyMaker/CommandTypeGenerator.cs:270]: his is the fix for #3114. The legacy Python loader defaults automate to false (genericcomps.py line 525). The previous C# code incorrectly defaulted to true, which broke Dynamo scripts saved in Manual run mode. This was already merged and verified

File: dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs:114]: Fixed under a new Commit

File: dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs [hunk 4]
```diff
@@ -79,6 +79,58 @@ private static void LogParseException(string parsedFile, Exception ex)
...
+            }
+
+            if (revitYear <= 0)
+            {
+                LogWarning("Skipping min / max version test, since Revit version is unknown");
+                return true;
+            }
+
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs]: Fixed under a new Commit

**Comment by tay0thman** [comment]
@devloai If you are available , review my responses to copilot AI

File: dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs [hunk 5]
```diff
@@ -90,6 +90,13 @@ private static void LogParseException(string parsedFile, Exception ex)
         /// <param name="name">The component or extension name, used in log messages.</param>
         private static bool IsRevitVersionCompatible(string minRevitVersion, string maxRevitVersion, int revitYear, string name)
         {
+            // Early exit when Revit version is unknown — skip version filtering entirely
+            if (revitYear <= 0)
+            {
+                LogWarning("Skipping min / max version test, since Revit version is unknown");
+                return true;
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitExtensionParser/ExtensionParser.cs:97]: This PR also includes a fix that passes the actual revitYear through ExtensionManagerService → ParseInstalledExtensions(), so revitYear <= 0 should no longer occur during normal Revit sessions. The early return is a defensive safety net for edge cases like unit tests or standalone parser usage outside of Revit. A log-once optimization can be addressed in a follow-up if needed.

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs [hunk 6]
```diff
@@ -129,58 +129,78 @@ public void LoadSession()
...
-                    // This matches the Python loader flow
-                    if (!string.IsNullOrEmpty(ext.StartupScript))
+                    _logger.Debug($"[PERF] {ext.Name} - Build: {buildTime}ms, Load: {stepStopwatch.ElapsedMilliseconds}ms");
+
+                    assembledExtensions.Add((ext, assmInfo));
+                }
+                catch (Exception ex)
+                {
+                    _logger.Error($"Error building/loading extension '{ext?.Name ?? "unknown"}': {ex.Message}");
+                }
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs:165]: updated all three catch blocks in this method to log ex (full ToString with stack trace) instead of ex.Message. Fixed in latest commit.

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs [hunk 7]
```diff
@@ -129,58 +129,78 @@ public void LoadSession()
...
                     {
                         _logger.Info($"Running startup tasks for {ext.Name}");
                         stepStopwatch.Restart();
                         ExecuteExtensionStartupScript(ext, libraryExtensions);
                         _logger.Debug($"[PERF] {ext.Name} - StartupScript: {stepStopwatch.ElapsedMilliseconds}ms");
                     }
+                    catch (Exception ex)
+                    {
+                        _logger.Error($"Startup script error for '{ext.Name}': {ex.Message}");
+                    }
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs:184]: updated all three catch blocks in this method to log ex (full ToString with stack trace) instead of ex.Message. Fixed in latest commit.

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs [hunk 8]
```diff
@@ -129,58 +129,78 @@ public void LoadSession()
...
                     stepStopwatch.Restart();
                     _uiManager?.BuildUI(ext, assmInfo);
                     _logger.Debug($"[PERF] {ext.Name} - BuildUI: {stepStopwatch.ElapsedMilliseconds}ms");
                     _logger.Info($"UI created for extension: {ext.Name}");
                 }
                 catch (Exception ex)
                 {
-                    _logger.Error($"Error processing extension '{ext?.Name ?? "unknown"}': {ex.Message}");
+                    _logger.Error($"UI build error for '{ext?.Name ?? "unknown"}': {ex.Message}");
                 }
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs:201]: updated all three catch blocks in this method to log ex (full ToString with stack trace) instead of ex.Message. Fixed in latest commit.

**Comment by tay0thman** [comment]
@jmcouffin  can you request another round of review by copilot? I think I covered all of its comments but just to make sure, I tested the DLLs locally and looks good here.

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/UIManager/SessionManagerService.cs:184]: Resolved in fix/csharp-loader-regressions new commmit
 4943c61c1..9157d7141  fix/csharp-loader-regressions

File: dev/pyRevitLoader/pyRevitAssemblyBuilder/SessionManager/ExtensionManagerService.cs [hunk 9]
```diff
@@ -2,7 +2,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using pyRevitExtensionParser;
-// using static pyRevitExtensionParser.ExtensionParser; // not currently used - all calls use fully qualified ExtensionParser.*
+using static pyRevitExtensionParser.ExtensionParser;
```
[the remark this answers is not in the document]
Comment (tay0thman) [reply dev/pyRevitLoader/pyRevitAssemblyBuilder/SessionManager/ExtensionManagerService.cs]: Resolved in fix/csharp-loader-regressions new commmit
 4943c61c1..9157d7141  fix/csharp-loader-regressions

**Comment by jmcouffin** [comment]
@romangolev can you take a look?

**Comment by romangolev** [comment]
@tay0thman @jmcouffin works well in Revit 2025  ^_^

**Comment by jmcouffin** [comment]
@tay0thman blessing from Roman!
Bravo!
